src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -2997,7 +2998,80 @@ private StatementAst ConfigurationStatementRule(IEnumerable<AttributeAst> custom
                         {
                             // Load the default CIM keywords
                             Collection<Exception> CIMKeywordErrors = new Collection<Exception>();
-                            Microsoft.PowerShell.DesiredStateConfiguration.Internal.DscClassCache.LoadDefaultCimKeywords(CIMKeywordErrors);
+                            if (ExperimentalFeature.IsEnabled("PSDscJsonSchemaSupport"))
+                            {
+                                // In addition to checking if experimental feature is enabled
+                                // also check if v3.0 (or later) of PSDesiredStateConfiguration module is available
+
+                                // pre-v3 module is mof-based
+                                // having a pre-v3 module pre-loaded gives user a way to force usage of mof-based APIs for dsc configuration compilation
+
+                                // First check if PSDesiredStateConfiguration is already loaded
+                                // if pre-v3 is already loaded then use old mof-based APIs
+                                // otherwise if v3.0.0 or later is already loaded then use new json-based APIS
+                                // otherwise no version of the module is currently loaded - then try to load v3 and use json-based APIs is loading is successful
+                                // if v3 can not be found/loaded then use old mof-based APIs
+
+                                p.AddCommand(new CmdletInfo("Get-Module", typeof(Microsoft.PowerShell.Commands.GetModuleCommand)));
+                                p.AddParameter("Name", "PSDesiredStateConfiguration");
+                                
+                                bool v3IsLoaded = false;
```

Comment:
This is pretty scary. Looking at the implementation today, the created PowerShell never gets used.

If we start running commands in the parser, then:

- Parser performance will go through the floor, which is bad on a number of levels
- Parser invocation can execute arbitrary code, which is a boundary we've previously not crossed. Parsing is currently considered safe
- Parsing will change machine state
- Interactive or repetitive parser invocations, like those in VSCode, PSReadLine or PSScriptAnalyzer, will all suffer when `configuration` blocks are present
- The test to get around this is also flawed because DynamicKeywords are thread-static. Invoking the parser from another thread will cause PowerShell creation and a module import, causing issues for parser-users that try to use thread pools for better throughput.

Reply:
/cc @daxian-dbw

Reply:
All my other comments in this PR are just suggestions, but I think this one is a blocker unfortunately

Reply:
This approach was the end result of discussion with Steve, Travis and Balu.  Not the most elegant solution, but for the lack of a better one, this gets things done.
I share the concern, however those negative side effects will appear only when `configuration` token is mentioned in the script.

Reply:
I agree with @rjmholt on this. For an application that only depends on `S.M.A` nuget package for parsing functionality, such as a script analyzer, this change would potentially break it because for a `PowerShell` instance created by `PowerShell.Create()`, it attempts to load all PS assemblies such as `Management.dll` and `Utility.dll` when calling `Invoke` on the `PowerShell` instance, but those assemblies are not available when it depends on `S.M.A` solely.

It looks to me there should be another experimental flag defined in the v3 PSDesiredConfiguration module. The `PSDscJsonSchemaSupport` experimental feature depends on the one defined in the module. So the code here should be
```
if (ExperimentalFeature.IsEnabled("PSDscJsonSchemaSupport"))
{
    if (!ExperimentalFeature.IsEnabled("PSDesiredConfiguration.DscJsonSchemaSupport"))
    {
        throw new NotSupportedException("The experimental feature 'PSDscJsonSchemaSupport' depends on the version 3 of the 'PSDesiredConfiguration' module, and the experimental feature 'PSDesiredConfiguration.DscJsonSchemaSupport' defined in that module should also be enabled.");
    }

    Internal.Json.DscClassCache.LoadDefaultCimKeywords(CIMKeywordErrors);
}
else
{
    Internal.DscClassCache.LoadDefaultCimKeywords(CIMKeywordErrors);
}
```
Is the v3 `PSDesiredConfiguration` backward compatible? No matter it's so or not, the new JSON feature should be presented as an experimental feature to start with.

That might be too much work to catch the preview.7 release, I suggest for now, just call `Internal.Json.DscClassCache.LoadDefaultCimKeywords(CIMKeywordErrors)` without checking for the existence of v3 `PSDesiredConfiguration`, and throw error in the method if the `*.schema.json` file cannot be found. In the exception message, say that v3 `PSDesiredConfiguration` is required.

The whole configuration related code looks pretty messy. We will have to refactor a bunch of them when separating the configuration parsing out along with the DSC stuff as a subsystem in 7.2.

Reply:
> but for the lack of a better one, this gets things done.

But is this the bar for changes to PowerShell, especially new features in the parser? We can't only accept perfect code, but if we accept everything that just works then we quickly assume technical debt and have no guarantee that it will ever be improved. Experimental feature or not, the code is in there and must be worked around.

But also as someone who works on a number of tools that trust a call like `Parser.ParseInput(...)` to be side-effect free, this just makes life incredibly difficult. Perhaps it only applies when the `configuration` keyword is present -- but how does my tool know that, since it can't parse the file to find out?

For an example of such a side effect:

```powershell
$psdscPath = "$PSHome/Modules/PSDesiredStateConfiguration"
mkdir $psdscPath/2.0.5
Move-Item $psdscPath/* $psdscPath/2.0.5
mkdir $psdscPath/3.0.0
New-ModuleManifest -Path $psdscPath/3.0.0/PSDesiredStateConfiguration -ModuleVersion 3.0.0 -ScriptsToProcess ./OnImport.ps1
New-Item -Path $psdscPath/3.0.0/OnImport.ps1 -Value @'
New-Item -Path ~/Desktop/dsc.txt -Value "Hello from the DSC module!"
'@

New-Item -Path ex.ps1 -Value "configuration MyConfiguration { }"
$toks = $errs = $null
[System.Management.Automation.Language.Parser]::ParseFile("$PWD/ex.ps1", [ref]$toks, [ref]$errs)
```

Reply:
I want to add:

> This approach was the end result of discussion with Steve, Travis and Balu

Given that this is a change in the parser that imports a module, I don't think you can conclude such a discussion without consulting owners of the parser and module code

Reply:
Some thoughts...

"an application that only depends on S.M.A nuget package for parsing functionality..." Ok, that makes sense.
"Is the v3 PSDesiredConfiguration backward compatible?" - no; v3 and beyond is expected to be working only with json schema.

"there should be another experimental flag defined in the v3 PSDesiredConfiguration module..." Sorry, I don't see how adding an additional exp flag in the module will help.... 
The fundamental problem that we have here in the parser code (and that was attempted to be solved with this module import) is 2 fold: 1) at parsing time we have to decide which APIs to use: mof-based or new json-based. The requirement is that by default existing mof-based should be used, additionally json-based should be used if exp.feature is enabled. However `configuration` syntax stays the same in both cases, so obvious choice is to use some sort of heuristic (like presence of v3 of DSC module on the system) or external switch (i.e. exp flag), especially considering next item 2) among other things, mof or json-based call to `LoadDefaultCimKeywords` essentially loads base schema classes - in the cross-platform context (aka. existing DSC module for linux) those come with the DSC module, that can theoretically be installed anywhere on the system (and this is partially why PS module load subsystem is used - to find module location on the system). The first thing DSC module does on loading is setting up DSC_HOME env var that points to itself (i.e. location of base classes schema), then this env var location is used by `LoadDefaultCimKeywords` to actually load base classes.
So basically it comes to this: how during `configuration` parsing can we know the location of DSC module on the system without utilizing PS module-loading subsystem?

"We will have to refactor a bunch of them when separating the configuration parsing out along with the DSC stuff as a subsystem" Definitely.

"do we have any tests for the DSC code?" Yes, there is some in this repo (that actually caught a couple of my mistakes in this parser changes), but majority is in DSC module repo.